nasa/harmony
# HARMONY-1315: Fairer queueing and redesign of polling for work and getting service metrics

## Jira Issue ID
HARMONY-1315

## Description
When polling for work we now look at the currently in progress work to determine the next user that should have work performed.

We also introduced the user_work table in this pull request to significantly improve performance of polling for work and capturing the number of items currently in work.

When there were hundreds of thousands of work items for a service the prior polling for work queries could take more than 90 seconds and now they take a few hundred milliseconds with the same amount of work items. Similarly the query to determine the number of work items ready or running for a service used by the HPA now takes milliseconds instead of tens of seconds with many work items.

## Local Test Steps
There were no changes to user facing functionality other than the fair queueing algorithm change. For testing we should regression test certain areas and verify the database migration.

1. Submit a few requests that will be in progress at the time of running the database migration. Pause a couple of the requests.
2. Run the database migration and verify the contents of the user_work table make sense. Verify any non-paused jobs work to completion after the migration and restarting harmony. Verify the paused jobs do not have work performed.
3. Resume one of the paused jobs and make sure it is worked to completion.
4. Cancel one of the jobs and make sure that the entries for that job are removed from the user_work table and there are no problems with polling.

Note that I also tested this branch in a sandbox environment.

## PR Acceptance Checklist
